airdrop/airdrop.gno

```diff
@@ -0,0 +1,215 @@
+	bitmap, exists := a.claimedBitmap[word]
+	if !exists {
+		return false
+	}
+	mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(index))
+	return new(u256.Uint).And(bitmap, mask).Cmp(u256.Zero()) != 0
+}
+
+// claimIDToBitmapIndex converts a claim ID to bitmap index
+func (a *Airdrop) claimIDToBitmapIndex(claimID uint64) (word uint64, index uint8) {
+	word = claimID / 256
+	index = uint8(claimID % 256)
+	return
+}
+
+// Claim processes a single claim
+func (a *Airdrop) Claim(claim Claim, proof [][]byte) (bool, error) {
+	valid, err := a.verifyProof(claim, proof)
+	if err != nil {
+		return false, err
+	}
+	if !valid {
+		return false, ErrInvalidProof
+	}
+
+	word, index := a.claimIDToBitmapIndex(claim.ID)
+	bitmap, exists := a.claimedBitmap[word]
+	if !exists {
+		bitmap = u256.Zero()
+	}
+	mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(index))
+	alreadyClaimed := new(u256.Uint).And(bitmap, mask).Cmp(u256.Zero()) != 0
+
+	if alreadyClaimed {
+		return false, nil // XXX should we return an error?
+	}
+
+	// update the bitmap
+	newBitmap := new(u256.Uint).Or(bitmap, mask)
+	a.claimedBitmap[word] = newBitmap
+
+	// transfer tokens
+	err = a.token.Transfer(std.Address(claim.Claimee), claim.Amount.Uint64())
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// verifyProof checks if the merkle proof is valid
+func (a *Airdrop) verifyProof(claim Claim, proff [][]byte) (bool, error) {
+	proof, err := a.merkleTree.Proof(claim)
+	if err != nil {
+		return false, err
+	}
+	return a.merkleTree.Verify(claim, proof), nil
+}
+
+// Claim128 processes up to 128 claims in a single batch
+func (a *Airdrop) Claim128(claims []Claim, remainingProof [][]byte) (uint8, error) {
+	if len(claims) == 0 || len(claims) > 128 {
+		return 0, ErrInvalidClaimsCount
+	}
+
+	// verify that the first claim ID is a multiple of 128
+	if claims[0].ID%128 != 0 {
+		return 0, ErrFirstClaimIsNot128
+	}
+
+	word := claims[0].ID / 256
+	bitmap, exists := a.claimedBitmap[word]
+	if !exists {
+		bitmap = u256.Zero()
+	}
+
+	var claimedCount uint8
+
+	for i, claim := range claims {
+		index := uint8(i)
+		mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(index))
+		alreadyClaimed := new(u256.Uint).And(bitmap, mask).Cmp(u256.Zero()) != 0
+
+		if !alreadyClaimed {
+			bitmap = new(u256.Uint).Or(bitmap, mask)
+			err := a.token.Transfer(std.Address(claim.Claimee), claim.Amount.Uint64())
+			if err != nil {
+				return claimedCount, err
+			}
+			claimedCount++
+		}
+	}
+
+	a.claimedBitmap[word] = bitmap
+	return claimedCount, nil
+}
+
+// func (a *Airdrop) Refund() error {
+// 	if a.config.RefundableTimestamp == 0 {
+// 		return ErrRefundableTimestampNotSet
+// 	}
+
+// 	currentTime := uint64(time.Now().Unix())
+// 	if currentTime < a.config.RefundableTimestamp {
+// 		return ErrTooEarlyToRefund
+// 	}
+
+// 	balance, err:= a.token.BalanceOf(a.GetToken()) // ?
+// 	if err != nil {
+// 		return err
+// 	}
+
+// 	return a.token.Transfer(a.config.RefundTo, balance)
+// }
```

Comment:
I couldn't find a function like getting the address of the token (i.e., `GetAddress`), so I commented it out for now.

Reply:
if you want handle multiple tokens, use register pattern